src/encoding.cc

```diff
@@ -271,3 +271,77 @@ double DecodeDouble(const char *ptr) {
   memcpy(&value, &decoded, sizeof(value));
   return value;
 }
+
+char* EncodeVarint32(char *dst, uint32_t v) {
```

Comment:
No logical mistake found. I just wonder if this loop unrolling will be more efficient, e.g. protobuf seems do not unroll it. https://sourcegraph.com/github.com/protocolbuffers/protobuf/-/blob/src/google/protobuf/io/coded_stream.h?L898:42&subtree=true

Reply:
I perform a benchmark here: https://quick-bench.com/q/NlktQnulDGxmP1pcIIK6MLvjldk

It seems that this unrolling do not take much performance improvement.

Maybe my benchmark has some flaw, so you guys can review it and correct me : )

Reply:
The benchmark looks good, NoUnroll even faster than Unroll in GCC.

Reply:
maybe many conditional judgments in unrolling break pipelining processing, this minor computation doesn't influence performance. And now @torwig implement new code, also is free from license problem.